File: src/main/kotlin/ninja/blacknet/time/SystemClock.kt
/*
 * Copyright (c) 2018-2020 Pavel Vasin
 *
 * Licensed under the Jelurida Public License version 1.1
 * for the Blacknet Public Blockchain Platform (the "License");
 * you may not use this file except in compliance with the License.
 * See the LICENSE.txt file at the top-level directory of this distribution.
 */

package ninja.blacknet.time

import ninja.blacknet.time.milliseconds.MilliSeconds

/**
 * Implementation of a clock that returns the current time of the operating system.
 */
object SystemClock {
    /**
     * The current time as the number of seconds since the Epoch.
     */
    val seconds: Long
        get() = System.currentTimeMillis() / 1000

    /**
     * The current time as the number of milliseconds since the Epoch.
     */
    val milliseconds: MilliSeconds
        get() = MilliSeconds(System.currentTimeMillis())
}

File: time/src/main/rust/systemclock.rs
/*
 * Copyright (c) 2025 Pavel Vasin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

use crate::milliseconds::Milliseconds;
use std::time::SystemTime;

// In milliseconds since UNIX epoch not counting leap seconds
pub struct SystemClock {}

impl SystemClock {
    pub fn now() -> Milliseconds {
        let now = SystemTime::now();
        match now.duration_since(SystemTime::UNIX_EPOCH) {
            Ok(since) => (since.as_millis() as i64).into(),
            Err(until) => (-(until.duration().as_millis() as i64)).into(),
        }
    }
}
